smtp.c

```diff
@@ -51,88 +59,170 @@ struct smtp_statistics {
 	struct ratelimitspp_statistics ratelimitspp;
 };
 
+struct smtp_statistics_vector {
+	struct vector maxconnnet;
+	struct vector maxconnip;
+	struct vector maxconnrule;
+	struct vector maxload;
+};
+
+struct smtp_statistics {
+	struct smtp_statistics_vector ssv;
+	struct smtp_statistics_scalar sss;
+};
+
 static
 struct
 ratelimitspp_statistics aggregated_ratelimtspp;
 
+static
+struct
+smtp_statistics_vector aggregated_limits;
+
 static
 void *
 smtp_data_init() {
 	struct smtp_statistics * ret;
 	ret = calloc(1, sizeof * ret);
+	vector_init(&ret->ssv.maxload, sizeof(struct limit_t));
+	vector_init(&ret->ssv.maxconnnet, sizeof(struct limit_t));
+	vector_init(&ret->ssv.maxconnip, sizeof(struct limit_t));
+	vector_init(&ret->ssv.maxconnrule, sizeof(struct limit_t));
+	vector_init(&aggregated_limits.maxload, sizeof(struct limit_t));
+	vector_init(&aggregated_limits.maxconnnet, sizeof(struct limit_t));
+	vector_init(&aggregated_limits.maxconnip, sizeof(struct limit_t));
+	vector_init(&aggregated_limits.maxconnrule, sizeof(struct limit_t));
 	return ret;
 }
 
+static
+void
+set_rulename(char * name_d, const char * name_s, const size_t size) {
+	memset(name_d, 0, size);
+	for (int i = 0; i < size; i++) {
+		if (!(name_s + i) || name_s[i] == ')') {
```

Comment:
You seem forget to dereference the pointer in the first condition., thus, you are not testing for presence of the null character. I suggest to fix it like this
```c
if (!name_s[i] || name_s[i] == ')')
```

Reply:
I would prefer
```c
if (name_s[i] == '\0' || name_s[i] == ')')
```
It is more readable to me.